### PR TITLE
Adds RFC for adding type hints to python scripts

### PR DESCRIPTION
**Pull request by mitchhentges**
[Rendered RFC](https://github.com/mitchhentges/releng-rfcs/blob/typing-python-scripts/rfcs/0012-typing-python-scripts.md)

### CODE REVIEW

File: rfcs/0012-typing-python-scripts.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# RFC 12 - Typing Python Scripts
+* Comments: [#12](https://api.github.com/repos/mozilla-releng/releng-rfcs/issues/12)
+* Proposed by: @mitchhentges
+
+# Summary
+
+[Typing hinting is in the Python standard library as of Python 3.5](https://docs.python.org/3/library/typing.html).
+We should slowly introduce it into our `*scripts` and `scriptworker`
```
[the remark this answers is not in the document]
Comment (mitchhentges) [reply rfcs/0012-typing-python-scripts.md]: Hmm, this isn't meant to limit the spread of typing.
I can lower-bound this: "We should slowly introduce it, starting with our `*scripts` and `scriptworker`". What do you think?

File: rfcs/0012-typing-python-scripts.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+* Improved IDE completion
+
+# Details
+
+Just as if introducing Typescript to a Javascript codebase (or Kotlin to Java), we can slowly add typings on a PR basis.
+Typed functions interact properly with untyped functions, so refactoring the entire codebase at once isn't necessary.
+
+[PEP 484 has a section on `*args` and `**kwargs`](https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values).
+
+Note: typings are ignored in production.
```
[the remark this answers is not in the document]
Comment (mitchhentges) [reply rfcs/0012-typing-python-scripts.md]: Yeah, Python's built-in `typings` module doesn't affect prod runtime: mismatched typings will fail silently in production.

I'd like to defer this as a separate discussion/RFC, since it will have further implications:
* We need to trust additional software at runtime - though `mypy`/etc may be relatively standard, production risk is non-zero
* There will be a performance impact. Hopefully small enough not to matter?

**Comment by mitchhentges** [comment]
Regarding rust: I'm very intrigued, but I think that migrating our scripts to Rust is a separate, much larger undertaking. I think that we should defer that to a separate discussion/RFC: if we're :+1: on it (I hope!), there's still significant value in adding types while we migrate over.

**Comment by escapewindow** [comment]
> 👍
> 
> Not to take away from this, but we should also consider whether it makes sense to use rust some of the places we currently use python.

+1, vague thoughts on rewriting scriptworker in go or rust. Rust is safer and potentially faster. Go is wider used and has an infra foothold in generic-worker and autograph. Also agreed it's a separate RFC and potentially much longer term :)

File: rfcs/0012-typing-python-scripts.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomprince) [reply rfcs/0012-typing-python-scripts.md]: All of these tools are static checkers that run in CI, instead of at runtime[1]. I'm not sure how much value there is in adding type annotation without something checking that they are or are not followed. I would be against adding type annotations without some form of tooling that used them.

[1] though there are also tools that do stuff at runtime

[the remark this answers is not in the document]
Comment (mitchhentges) [reply rfcs/0012-typing-python-scripts.md]: A primary benefit that I'm selfishly looking for is improving the usefulness of my IDE, since it'll be able to use type hints to give me more relevant code completions, etc.

I'm definitely interested in having a tool like `mypy` verify our code as part of CI, I'd just want to be able to configure it such that the build will still pass if we have untyped code.

[the remark this answers is not in the document]
Comment (escapewindow) [reply rfcs/0012-typing-python-scripts.md]: If we want to uniformly adjust all the *scripts' CI, that may be easier to do with a [monorepo](https://github.com/mozilla-releng/releng-rfcs/pull/8) ;-)

[the remark this answers is not in the document]
Comment (tomprince) [reply rfcs/0012-typing-python-scripts.md]: I haven't used any of the tools I listed, but my understanding (which matches my quick check of the documentation) is that they all support gradual typing, so that they will check what they can, but will still allow untyped code.

File: rfcs/0012-typing-python-scripts.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (garbas) [reply rfcs/0012-typing-python-scripts.md]: we already started https://github.com/mozilla/release-services/blob/master/src/shipit/api/shipit_api/product_details.py, but adding type signatures takes time and I'm not sure we have it (I'd like to be wrong :wink:). I'll leave my comment on the python typing system at the bottom.

**Comment by garbas** [comment]
on typing and mypy: very primitive (if you ever used any other type system) even compared to typescript. probably primitive is not the right word, maybe *less advance then other type systems* would be better. i don't know i'm slow with finding the right words for the *emotions* at this time.

also adding types in python will not make your system easy to refactor since this is gradually type system and we can not really trust the compiler/linter. it would uncover few bugs and maybe make few nicer suggestions in your IDE and maybe even make it a little easier. but this all also comes with a cost of writing type hints and it is not pleasant experience. At least I didn't enjoy it that much when writing [product details](https://github.com/mozilla/release-services/blob/master/src/shipit/api/shipit_api/product_details.py).

error messages are poor, I encounter many limitation, there are corner cases you need to know about, not all latest python features are supported and also I found out 2 bugs in mypy. All in all very bad user experience. it really felt like fighting the compiler/linter most of the time.

i think the underlying motivation for this RFC is improve maintenance in the long term and i'm not sure using type hints is the solution. i found structuring the code to isolate IO to be much more effective and less time consuming to learn.

now i'm not against type hints, but i would rather see developer decide at the time of writing code if adding type hints would benefit the project.

I'm sure mypy and typing will improve overtime, but i'm not sure i'm (we're) willing to wait. i'd rather look into using rust for our future stuff or future rewrites of our existing tools.

**Comment by mitchhentges** [comment]
Thanks garbas!
I'm a little bummed that the type analysis tooling isn't quite as mature just yet.

Even if we don't decide to start a gradual, long-term refactoring to add type hints, I believe that there's still benefit in at least allowing type hints to exist in the codebase. For example, I ran into a situation earlier where I was implementing a function that took some primitives (some `str` and some `int`). Without type hints, some of these variables were hard to keep track of (is my "version" argument a semver string? Or just an incrementing integer?). In cases like that, adding type hints would've made integrating that function a little more convenient, especially with my IDE's help.

Perhaps one of the reasons that adding all the typings to [product_details.py](https://github.com/mozilla/release-services/blob/master/src/shipit/api/shipit_api/product_details.py) was so costly was because typings were added uniformly, despite varying benefit for each type? I'd for this RFC to represent adding typings organically to areas that would benefit the most.

**Comment by mitchhentges** [comment]
@garbas how recently did you run into the instability with `mypy`? Was this within the last year?

**Comment by garbas** [comment]
> I'm a little bummed that the type analysis tooling isn't quite as mature just yet.

yeah I had high hopes as well

> Even if we don't decide to start a gradual, long-term refactoring to add type hints, I believe that there's still benefit in at least allowing type hints to exist in the codebase. For example, I ran into a situation earlier where I was implementing a function that took some primitives (some str and some int). Without type hints, some of these variables were hard to keep track of (is my "version" argument a semver string? Or just an incrementing integer?). In cases like that, adding type hints would've made integrating that function a little more convenient, especially with my IDE's help.

For simple functions this is not a problem, but very soon you will run into custom types and type structured. And then you quickly start fighting the compiler/linter.

> Perhaps one of the reasons that adding all the typings to product_details.py was so costly was because typings were added uniformly, despite varying benefit for each type? I'd for this RFC to represent adding typings organically to areas that would benefit the most.

product details was written from scratch and as you should we started with types and modeling our data with it. i did this exercise many times in haskell, elm and typescript  and python tooling and current support for it was really poor. 

in release services we already enabled mypy as part of the flake8 checks (flake8-mypy). so whenever somebody starts adding type hints we are going to check it. but we are not enforcing everybody to use it.

> @garbas how recently did you run into the instability with `mypy`? Was this within the last year?

We gave mypy a try in Q4 last year.

File: rfcs/0012-typing-python-scripts.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Callek) [reply rfcs/0012-typing-python-scripts.md]: I am firmly +1 on the feeling that if we do this we should use mypy to enforce the type hints we use.

File: rfcs/0012-typing-python-scripts.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+# Open Questions
+
+* Should we using tooling such as `mypy` in CI to statically analyze our typings?
+    * Suggestion: perhaps we could add it to CI, but not let it fail the build? Watch this closely, and re-visit whether it's useful in a month or two?
```
Comment (Callek) [inline rfcs/0012-typing-python-scripts.md]: I think we should start with a few basic checks with mypy, and *exclude* test directories.

the mindset to me is that we want to allow invalid type passing in tests, because unless we enforce typing using some tool we'll still have potential for invalid input, and should have clearly defined output in invalid input cases.

We also should (arguably) in library type code have a set of tests that exist to example the typed functions so we know that the types we set are correct and that we don't (unintentionally) regress them. This would be for cases of expecting to pass a tuple of strings to a scriptworker function, but instead we send file-objects and stuff breaks. etc.

Lastly there was an expiriment I did with typing+mypy a while back using mozilla-version and at that time I didn't think it was worth following through on, of course I went as far as typing against the full tests, supporting py2, and using Strict mode.  https://github.com/mozilla-releng/mozilla-version/pull/32

Comment (Callek) [reply rfcs/0012-typing-python-scripts.md]: The primary benefit of using mypy is additionally that in typed functions we can better-vet package updates for our depended-on packages when they also provide typehints, e.g. if an API changes to accept something different in an area of code we use, we will get a better sense of "this is likely wrong" than we would today (especially if we don't fully vet all upstream things) -- this is doubly useful in cases like:

* Releng Package --> Upstream Package A --> Upstream Package B

Where A depends on B, B gets bumped, A uses B but we use A, and the behavior we expect in A gets broken by B but we mock the behavior in A instead of exercising it in tests

Comment (mitchhentges) [reply rfcs/0012-typing-python-scripts.md]: I'm mostly on the same page here, though some thoughts on:

> the mindset to me is that we want to allow invalid type passing in tests, because unless we enforce typing using some tool we'll still have potential for invalid input, and should have clearly defined output in invalid input cases.

I'm not sure that tests that check "if a function accepts two `str` but I give it two `dict`, it should..." are necessarily valuable for two reasons:
* They can easily become a rabbit hole (what about `list, list`? `int, dict`? `None, None`?)
* Tests that check validation for incorrect user input is a big :+1: from me, but if we're just verifying an API, I wouldn't get too invested in validation, because there's only so much that you can do.
    * For example, you have a function that takes a string email address. You _can_ have validation for incorrect types, the string missing the `@`, or for checking the thousands of ways an email address can be formatted - though, at the end of the day, the caller can _still_ succeed validation but give you a wrong email. For these cases, it might be easier to just lean on "garbage in, garbage out" and not spend too much energy trying to validate non-user-input

-----

I got a bit distracted/pedantic, but TL;DR: I'm on the same page, I agree that tests shouldn't be `mypy`'d because they're usually a little more flexible with types (e.g.: how will `MagicMock` behave?). However, I'm less on-board with tests that assert that we hand-hold consumers if types are invalid.

File: rfcs/0012-typing-python-scripts.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mitchhentges) [reply rfcs/0012-typing-python-scripts.md]: Perhaps a good solution that will address all concerns is:

* Allow typings
    * Don't require all new code to be typed! Types can be added on a as-needed basis
* Set up with `mypy`
    * Enforce `mypy` checks pass before PRs are allowed to be merged
    * If a certain function signature isn't working nicely with `mypy`, we omit the typings from that function
    * If `mypy` is behaving intermittently or harming our process, we can re-visit if we should remove it

We can start `mypy` enforcement with with a subset of projects (perhaps `scriptworker`? Maybe something less commonly used, like `pushapkscript`?). If it's :100:, we can roll out our `mypy` enforcement to all of our *scripts, and potentially eventually all releng python code.

**Comment by mitchhentges** [comment]
@garbas I've modified this RFC to (ideally) reflect and resolve your concerns?
TL;DR: I don't think that we should add type hints to _everything_ in the project. I think that we should just add `mypy` coverage on-PR, then allow devs to add type hints as-needed. I imagine that this will manifest as little pockets of type hints appearing on our codebases that slowly spread. When we run into `mypy` issues, we just avoid the problematic hints.

What do you think?

**Comment by garbas** [comment]
@mitchhentges until we rewrite everything in rust :smile: hehe

**Comment by mitchhentges** [comment]
:crab: Don't tempt me :smile: :crab:

**Comment by mitchhentges** [comment]
Bumping to "Final Comment" since there aren't any open questions, and things have settled down
